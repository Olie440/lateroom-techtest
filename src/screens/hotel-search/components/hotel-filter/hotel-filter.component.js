import React, { Component } from 'react';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faFilter } from '@fortawesome/free-solid-svg-icons';
import { toggleFilter } from '../../../../redux/actions/filter-hotels';

import './hotel-filter.css';

export class HotelFilter extends Component {
    state = {
        open: false
    }

    render() {
        const buttonDisabled = this.props.availableFilters.length === 0;

        return (
            <div className="hotel-filter">
                <button className="hotel-filter__button" onClick={this.toggleOpen} disabled={buttonDisabled}>
                    <FontAwesomeIcon icon={faFilter} />
                </button>
                { this.state.open && this.renderMenu() }
            </div>
        )
    }

    renderMenu() {
        return (
            <div className="hotel-filter__menu">
                { this.props.availableFilters.map(this.renderMenuItem) }
            </div>
        )
    }

    renderMenuItem = (item) => {
        const checked = this.props.appliedFilters.includes(item);

        return (
            <div className="hotel-filter__menu-item" key={item}>
                <input
                    type="checkbox"
                    checked={checked}
                    value={item}
                    onChange={this.toggleFilter} />
                { item }
            </div>
        )
    }

    toggleOpen = () => {
        this.setState((state) => ({
            open: !state.open
        }));
    }

    toggleFilter = (event) => {
        const { value, checked } = event.target;
        this.props.toggleFilter(value, checked);
    }
}

export function mapStateToProps(state) {
    return {
        ...state.filters
    }
}

const actions = {
    toggleFilter
};

export default connect(mapStateToProps, actions)(HotelFilter);
