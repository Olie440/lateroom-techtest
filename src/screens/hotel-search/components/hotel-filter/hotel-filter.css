.hotel-filter {
    position: relative;
    text-align: right;
}

.hotel-filter__menu {
    position: absolute;
    right: 0;
    text-align: left;
    padding: 1rem;
    background: #ffffff;
    border: 1px solid #cccccc;
}

.hotel-filter__menu-item {
    height: 1.5rem;
}

.hotel-filter__menu-item input {
    vertical-align: sub;
    margin-right: 0.25em;
}
