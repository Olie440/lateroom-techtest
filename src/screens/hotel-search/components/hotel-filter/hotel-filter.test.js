import React from 'react';
import { HotelFilter, mapStateToProps } from './hotel-filter.component';
import { shallow } from 'enzyme';

describe('HotelFilter', () => {
    let component, props;

    beforeEach(() => {
        props = {
            appliedFilters: ['car park'],
            availableFilters: ['car park', 'pool', 'gym'],
            toggleFilter: jest.fn()
        }
        component = shallow(<HotelFilter {...props} />);
    });

    describe('when state.open = false', () => {
        it('only renders the filter button', () => {
            expect(component).toMatchSnapshot();
        });

        it('toggles state.open = true when clicked', () => {
            expect(component.state('open')).toEqual(false);
            component.find('.hotel-filter__button').simulate('click');
            expect(component.state('open')).toEqual(true);
        });
    });

    describe('when state.open = true', () => {
        beforeEach(() => {
            component.setState({ open: true });
        });

        it('only renders the filter button', () => {
            expect(component).toMatchSnapshot();
        });

        it('toggles state.open = false when clicked', () => {
            expect(component.state('open')).toEqual(true);
            component.find('.hotel-filter__button').simulate('click');
            expect(component.state('open')).toEqual(false);
        });

        it('calls toggleFilter with the name and false when a checked filter is clicked', () => {
            const fakeEvent = {
                target: {
                    value: 'car park',
                    checked: false
                }
            };
            component.find('input').at(0).simulate('change', fakeEvent);
            expect(props.toggleFilter).toHaveBeenCalledWith('car park', false);
        });

        it('calls toggleFilter with the name and true when a unchecked filter is clicked', () => {
            const fakeEvent = {
                target: {
                    value: 'car park',
                    checked: true
                }
            };
            component.find('input').at(0).simulate('change', fakeEvent);
            expect(props.toggleFilter).toHaveBeenCalledWith('car park', true);
        });
    });

    it('disables the button when availableFilters.length = 0', () => {
        component.setProps({
            availableFilters: []
        });

        expect(component).toMatchSnapshot();
    });

    describe('mapStateToProps', () => {
        it('destructures the filters key', () => {
            const result = mapStateToProps({
                filters: {
                    availableFilters: ['foo'],
                    appliedFilters: []
                }
            });

            expect(result).toEqual({
                availableFilters: ['foo'],
                appliedFilters: []
            });
        })
    })
})
