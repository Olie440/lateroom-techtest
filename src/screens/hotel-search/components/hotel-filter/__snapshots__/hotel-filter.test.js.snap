// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelFilter disables the button when availableFilters.length = 0 1`] = `
<div
  className="hotel-filter"
>
  <button
    className="hotel-filter__button"
    disabled={true}
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f0b0",
            "M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z",
          ],
          "iconName": "filter",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
</div>
`;

exports[`HotelFilter when state.open = false only renders the filter button 1`] = `
<div
  className="hotel-filter"
>
  <button
    className="hotel-filter__button"
    disabled={false}
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f0b0",
            "M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z",
          ],
          "iconName": "filter",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
</div>
`;

exports[`HotelFilter when state.open = true only renders the filter button 1`] = `
<div
  className="hotel-filter"
>
  <button
    className="hotel-filter__button"
    disabled={false}
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f0b0",
            "M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z",
          ],
          "iconName": "filter",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
  <div
    className="hotel-filter__menu"
  >
    <div
      className="hotel-filter__menu-item"
      key="car park"
    >
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
        value="car park"
      />
      car park
    </div>
    <div
      className="hotel-filter__menu-item"
      key="pool"
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
        value="pool"
      />
      pool
    </div>
    <div
      className="hotel-filter__menu-item"
      key="gym"
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
        value="gym"
      />
      gym
    </div>
  </div>
</div>
`;
